src/coredump/coredump-client.c

```diff
@@ -0,0 +1,117 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "sd-event.h"
+
+#include "coredump-client.h"
+#include "coredump-config.h"
+#include "coredump-context.h"
+#include "coredump-socket.h"
+#include "coredump-submit.h"
+#include "coredump-util.h"
+#include "fd-util.h"
+#include "log.h"
+#include "namespace-util.h"
+#include "time-util.h"
+#include "varlink-io.systemd.Coredump.Client.h"
+#include "varlink-util.h"
+
+static int vl_method_submit_coredump(sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        CoredumpContext *context = ASSERT_PTR(userdata);
+        struct {
+                unsigned coredump_fd_index;
+                usec_t timestamp;
+                bool request_mode;
+        } p = {};
+        int r;
+
+        assert(link);
+
+        static const sd_json_dispatch_field dispatch_table[] = {
+                { "coredumpFileDescriptor", _SD_JSON_VARIANT_TYPE_INVALID, sd_json_dispatch_uint,    voffsetof(p, coredump_fd_index),  SD_JSON_MANDATORY },
+                { "timestamp",              _SD_JSON_VARIANT_TYPE_INVALID, sd_json_dispatch_uint64,  voffsetof(p, timestamp),          0                 },
+                { "requestMode",            SD_JSON_VARIANT_BOOLEAN,       sd_json_dispatch_stdbool, voffsetof(p, request_mode),       SD_JSON_MANDATORY },
+                {}
+        };
+
+        r = sd_varlink_dispatch(link, parameters, dispatch_table, &p);
+        if (r != 0)
+                return r;
+
+        context->request_mode = p.request_mode;
+        context->forwarded = true;
+
+        /* If we are in a non-initial time namespace, ignore the received timestamp. */
+        if (!timestamp_is_set(p.timestamp) || namespace_is_init(NAMESPACE_TIME) == 0) {
+                r = sd_event_now(sd_varlink_get_event(link), CLOCK_REALTIME, &context->timestamp);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to get the current time: %m");
+        } else
+                context->timestamp = p.timestamp;
+
+        r = sd_varlink_take_fd(link, p.coredump_fd_index);
+        if (r < 0)
+                return log_error_errno(r, "Failed to take file descriptor of the coredump socket: %m");
+        context->input_fd = r;
+
+        return sd_varlink_reply(link, /* parameters= */ NULL);
+}
+
+int coredump_client(int argc, char *argv[]) {
+        int r;
+
+        log_parse_environment();
+        log_set_target_and_open(LOG_TARGET_KMSG);
+
+        /* Make sure we never enter a loop. */
+        (void) set_dumpable(SUID_DUMP_DISABLE);
+
+        /* Ignore all parse errors. */
+        CoredumpConfig config = COREDUMP_CONFIG_NULL;
+        (void) coredump_parse_config(&config);
+
+        _cleanup_(coredump_context_done) CoredumpContext context = COREDUMP_CONTEXT_NULL;
+
+        _cleanup_(sd_varlink_server_unrefp) sd_varlink_server *varlink_server = NULL;
+        r = varlink_server_new(
```

Comment:
hmm, so we are in a function called "client", in a part of a program that we end up in if the binary is called with "--client" on its cmdline, but the primary thing this does, is allocate a *server*. That is quite confusing.

I think we should maybe change names here, i.e. leave client/server for the roles of the two sides in an IPC context. And maybe use Submitter/Receiver for the two roles when it comes to coredumps, to make clear which direction the coredump flows.